mkdocs/structure/pages.py

```diff
@@ -0,0 +1,143 @@
+from mkdocs.compat import urlparse, urlunparse
+from mkdocs.structure.toc import get_toc
+from mkdocs.structure.urls import get_relative_url
+
+from markdown.extensions import Extension
+from markdown.treeprocessors import Treeprocessor
+from markdown.util import AMP_SUBSTITUTE
+import markdown
+import io
+import os
+
+
+class Page(object):
+    def __init__(self, title, filepath):
+        self.title = title
+        self.file = None
+
+        # Navigation attributes
+        self.parent = None
+        self.previous = None
+        self.next = None
+
+        self.is_section = False
+        self.is_page = True
+
+        # Build attributes
+        self.markdown = None
+        self.html = None
+        self.meta = {}
+        self.toc = []
+
+        # '_filepath' is only used temporarily, when a page is first
+        # created based on the 'pages' configuration.
+        # The '.file' attribute should be used as the public API.
+        self._filepath = filepath
+
+    def build(self, md):
+        input_filepath = self.file.full_input_path
+        self.markdown = io.open(input_filepath, 'r', encoding='utf-8').read()
+        self.html = md.convert(self.markdown)
+        self.meta = getattr(md, 'Meta', {})
+        self.toc = get_toc(getattr(md, 'toc', ''))
+        if self.title is None:
+            self.title = self.get_default_title()
+
+    def get_default_title(self):
+        # Determine the title based on the first item in the table of contents.
+        if self.toc:
+            return self.toc[0].title
+
+        # Failing that, determine the title based on the filename.
+        title = self.file.root.replace('-', ' ').replace('_', ' ')
+        if title.lower() == title:
+            # Capitalize if the filename was all lowercase.
+            title = title.capitalize()
+        return title
+
+
+def build_pages(files):
+    pages = []
+    for file in files.documentation_pages():
+        md = markdown.Markdown(
+            extensions=[
+                _RelativePathExtension(file, files, strict=False),
+                'meta', 'toc', 'tables', 'fenced_code'
```

Comment:
Also, the default extensions are defined in the config. That way, duplicates are handled properly, etc (a number of people have reported setting the permalink config setting on the toc extension (including the MkdDocs docs; see [here][1]), which would define it a second time. The current solution handles that properly).  The config would need to be passed to this build_pages function. Or at least both the `markdown_extensions` key and the (private) `mdx_configs` key (which holds extension configs).

[1]: https://github.com/mkdocs/mkdocs/blob/master/mkdocs.yml#L23